src/pudl/output/ferc1.py

```diff
@@ -1063,6 +1077,88 @@ def calculations(self):
+          is a leaf, then it will have a single null calculation component)
+        - There should be no records where only one of table_name or xbrl_factoid
+          are null. They should either both or neither be null.
+        - table_name_parent and xbrl_factoid_parent should be non-null.
+        """
+        calc_cols = list(NodeId._fields)
+        parent_cols = [col + "_parent" for col in calc_cols]
+
+        # Keep only records where parent is within exploded tables:
+        calc_explode = self.calculation_components_xbrl_ferc1[
+            self.calculation_components_xbrl_ferc1.table_name_parent.isin(
+                self.table_names
+            )
+        ].copy()
+        # Groupby parent factoids
+        gb = calc_explode.groupby(
+            ["table_name_parent", "xbrl_factoid_parent"], as_index=False
+        )
+        # Identify groups where all calculation components are within the explosion
+        # We need to include table_name of pd.NA to retain non-calculated values.
+        calc_explode["is_within_explosion"] = gb.table_name.transform(
+            lambda x: x.isin(self.table_names + [pd.NA]).all()
+        )
+        # Convert any calculation componets from outside the explosion into reported
+        # values / leaf nodes:
+        calc_explode.loc[
+            ~calc_explode.is_within_explosion, list(NodeId._fields) + ["weight"]
+        ] = pd.NA
+        calc_explode = calc_explode.drop(columns=["is_within_explosion"]).reset_index(
+            drop=True
+        )
+
+        ##############################################################################
+        # Everything below here is error checking / debugging
+
+        # There should be no cases where only one of table_name or xbrl_factoid is NA:
+        partially_null = calc_explode[
+            calc_explode.table_name.isna() ^ calc_explode.xbrl_factoid.isna()
+        ]
+        if not partially_null.empty:
+            logger.error(
+                "Found unexpectedly null calculation components:\n"
+                f"{partially_null[['table_name', 'xbrl_factoid']]}"
+            )
+
+        # Using list of NodeIds rather than index here due to issues with NA types and
+        # it also avoids needing to rename columns or index level names.
+        # An index of all non-null child nodes / calculation components:
+        calc_idx = [
+            NodeId(*x)
+            for x in calc_explode[calc_cols].dropna(how="all").itertuples(index=False)
+        ]
+        # An index of all parent nodes:
+        parent_idx = [
+            NodeId(*x)
+            for x in calc_explode[parent_cols].dropna(how="all").itertuples(index=False)
+        ]
+
+        # Nodes that *only* appear as children / calculation components:
+        calc_only = list(set(calc_idx).difference(parent_idx))
+        if len(calc_only) > 0:
+            logger.error(
+                f"Found {len(calc_only)} nodes which *only* appear as calc components."
+            )
```

Comment:
spot checking the calc-only nodes it looks like many of them are the total -> subdimension records.